### PR TITLE
Check for client creds

### PR DESCRIPTION
**Pull request by tvh**

### CODE REVIEW

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs [hunk 1]
```diff
@@ -44,7 +49,16 @@ authorizeEndpoint = writeText "U AM U?"
 tokenEndpoint
     :: Handler b (OAuth2 IO b) ()
 tokenEndpoint = do
-    let grant_type = GrantPassword
+    clientAuth
+    req <- getRequest
+    let grant_type'' = S.lookup "grant_type" (headers req)
```
Comment (thsutton) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: Isn't `grant_type` a form encoded variable in the request body? And not a header.

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs [hunk 2]
```diff
@@ -44,7 +49,16 @@ authorizeEndpoint = writeText "U AM U?"
...
+    grant_type' <- case grant_type'' of
+        Just [grant_type'] -> return $ T.decodeUtf8 grant_type'
+        _ -> do
+            modifyResponse $ setResponseStatus 400 "grant_type missing"
+            r <- getResponse
+            finishWith r
+    let grant_type = grantType grant_type'
     case grant_type of
```
Comment (thsutton) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: No need for another variable `grant_type`

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs [hunk 3]
```diff
@@ -64,6 +104,15 @@ createGrant
     :: Handler b (OAuth2 IO b) TokenGrant
 createGrant = return aGrant
 
+-- | Create an access token and send it to the client.
+createAndServeToken
```
Comment (thsutton) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: Should probably be called `clientGrant`. Would be good to have a link to the RFC section which describes it too.

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs (hunk 1 above)
Comment (tvh) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: Yes, it is... fixing it up

File: oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs (hunk 3 above)
Comment (tvh) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: This should be called `createAndServeToken`, as it is exactly the same procedure for every single `grant_type`. It shouldn't be called directly though.

Comment (thsutton) [inline oauth2-server-snap/lib/Network/OAuth2/Server/Snap.hs]: We need a `GrantRequest` type for this to take as a param, encompassing type, scope, etc.

**Pull request merged**